CHANGELOG.md

```diff
@@ -4,6 +4,120 @@ All notable changes to Router will be documented in this file.
+
+- Improves the heuristics used to try to reuse the query named fragments in subgraph fetches. Said fragment will be reused ([apollographql/federation#2541](https://github.com/apollographql/federation/pull/2541)) more often, which can lead to smaller subgraph queries (and hence overall faster processing).
+- Fix potential assertion error during query planning in some multi-field `@requires` case. This error could be triggered ([#2575](https://github.com/apollographql/federation/pull/2575)) when a field in a `@requires` depended on another field that was also part of that same requires (for instance, if a field has a `@requires(fields: "id otherField")` and that `id` is also a key necessary to reach the subgraph providing `otherField`).
+  
+  The assertion error thrown in that case contained the message `Root groups (...) should have no remaining groups unhandled (...)`
+
+By [@abernix](https://github.com/abernix) in https://github.com/apollographql/router/pull/3107
+
+### Add support for throwing graphql errors in rhai responses ([Issue #3069](https://github.com/apollographql/router/issues/3069))
+
+It's possible to throw a graphql error from rhai when processing a request. This extends the capability to include when processing a response.
+
+Refer to the `Terminating client requests` section of the [Rhai api documentation](https://www.apollographql.com/docs/router/configuration/rhai) to learn how to throw GraphQL payloads.
+
+By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/3089
+
+### use a parking-lot mutex in Context ([Issue #2751](https://github.com/apollographql/router/issues/2751))
+
+The context requires synchronized access to the busy timer, and precedently we used a futures aware mutex for that, but those are susceptible to contention. This replaces that mutex with a parking-lot synchronous mutex that is much faster.
+
+By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/2885
+
+### Config and schema reloads now use async IO ([Issue #2613](https://github.com/apollographql/router/issues/2613))
+
+If you were using local schema or config then previously the Router was performing blocking IO in an async thread. This could have caused stalls to serving requests and was generally bad practice.
+The Router now uses async IO for all config and schema reloads.
+
+Fixing the above surfaced an issue with the experimental `force_hot_reload` feature introduced for testing. This has also been fixed and renamed to `force_reload`. 
+
+```diff
+experimental_chaos:
+-    force_hot_reload: 1m
++    force_reload: 1m
+```
+
+By [@bryncooke](https://github.com/bryncooke) in https://github.com/apollographql/router/pull/3016
+
+### Improve subgraph co-processor context processing ([Issue #3058](https://github.com/apollographql/router/issues/3058))
```

Comment:
We need to be consistent with what we call it - co-processor or coprocessor?

Reply:
[coprocessor](https://www.apollographql.com/docs/router/customizations/coprocessor/)! Updated — good catch. 😄